E2E Go testsuite

mgencur opened the pull request:
Fixes #411 

* A lot of the code comes from updated dependencies or from copying files over from knative/serving. You'll see those in separate commits.
* I've tested this against GKE, uploading images to GCR. I'm in trouble running this on Minikube as I'm still waiting on a new laptop with more RAM and the current laptop doesn't manage running knative/eventing with Minikube:). There might be some changes required for Minikube.
* The file called adding_tests.md is not copied over from knative/serving. Let me know if we need it here as well or if we need more documentation.
* Some source code that was copied from knative/serving is not used right now but it might be good to keep it in this repo for future use (but there's not much of it)
* The TestKubernetesEvents test configures various component in the test body. We might move it to utility classes later when we have the need to re-use some of the code. Please let me know what is preferred here.
* The original test included creation of Channel, Subscription and also Flow. I did not include the first two in the new test because the Flow will create them automatically (there were some redundant resources deployed with the original test which were unused)

## Proposed Changes

  * re-implement the bash-like tests in Go
  * implement various helper functions
  * implement proper cleanup which removes resources in reverse order of their creation
  * the test images are published before the test (similar to knative/serving's approach) and the test only references them
  * the original test yaml files are now removed

mgencur commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,289 @@
+
+// CreateRouteAndConfig will create Route and Config objects using clients.
+// The Config object will serve requests to a container started from the image at imagePath.
+func CreateRouteAndConfig(clients *test.Clients, logger *zap.SugaredLogger, cleaner *test.Cleaner, name string, imagePath string) error {
```
It is similar but slightly different.

adrcunha replied:
@srinivashegde86 can we make this flexible and have a single instance in knative/pkg?

mgencur replied:
There are two reasons why this is now in the eventing repo.
1) I'm using the `Cleaner` and adding the Route and Config to this Cleaner so that is can be later deleted. The point is deleting the objects in reverse order. This is currently not done in knative/serving and I had problems when deleting some resources because a ServiceAccount was deleted earlier and Feed/EventType could never be deleted afterwards. So maybe we could move the Cleaner to `knative/pkg` and use it in `knative/serving` as well
2) it is more flexible to leave creation of some objects in the individual repos. Having this in `pkg` means they have to be more general. And if they're not general enough in the future they will require changes in 3 repos (pkg, serving, eventing) rather than just one. The objects are too specific to be moved to pkq, IMO. My 2c

adrcunha replied:
I think [1] is a very compelling argument. Can you please open an issue?
I get your point in [2], I was suggesting just moving the common stuff to `pkg`.
But filing the issue would be enough, so I see no reason to block this PR.
Thanks!

mgencur replied:
Cool, created https://github.com/knative/eventing/issues/431

mgencur commented on test/upload-test-images.sh:
```diff
@@ -0,0 +1,33 @@
+IMAGE_PATHS_FILE="$(find $(dirname $0)/image_paths.txt)"
+DOCKER_TAG=$1
+
+while IFS='' read -r IMAGE || [[ -n "$IMAGE" ]]; do
```
Thanks, removing IFS. The `-n` parameter makes the script slightly more robust. The last line in image_paths.txt doesn't need to end with a new line character.

mgencur commented on test/upload-test-images.sh:
```diff
@@ -0,0 +1,33 @@
+: ${DOCKER_REPO_OVERRIDE:?"You must set 'DOCKER_REPO_OVERRIDE', see DEVELOPMENT.md"}
+
+export KO_DOCKER_REPO=${DOCKER_REPO_OVERRIDE}
+IMAGE_PATHS_FILE="$(find $(dirname $0)/image_paths.txt)"
```
Using `find` is more robust in the way that the script can be run from both eventing/ and eventing/test folder. Not a big deal but I think it's useful. Using just relative path means I can only run it from eventing/test folder.

adrcunha replied:
If you're using `dirname $0` the script will also be runnable from eventing/ and from eventing/test, since its location will be relative to this script.

mgencur replied:
Right.

mgencur commented:
@adrcunha , I think all the comments are addressed except for the readme file. Can you let me know if this is really required? Please see my question above.

adrcunha commented on test/README.md:
```diff
@@ -14,26 +15,83 @@ Use `go test`:
+
+Simply run the `./test/e2e-tests.sh` script. It will create a GKE cluster, install Knative Serving stack with Istio, upload test images to your Docker repo and run the end-to-end tests against the Knative Eventing built from source.
+
+If you already have the `*_OVERRIDE` environment variables set, call the script with the `--run-tests arguments` and it will use the cluster and run the tests. Note that this requires you to have Serving and Istio installed and configured to your particular configuration setup. Knative Eventing will still built and deployed from source.
```
mega nit: "arguments" should be "argument".

adrcunha commented on test/crd_checks.go:
```diff
@@ -0,0 +1,212 @@
+/*
```
Looks like you have mismatches due to different versions of the code? Did you try updating the knative/pkg dependencies? @srinivashegde86 might be able to help.

mgencur replied:
Thanks. I've already solved this problem.

adrcunha commented on test/image_paths.txt:
```diff
@@ -0,0 +1,4 @@
+github.com/knative/eventing/pkg/buses/stub/dispatcher
```
No comments on this one?

mgencur replied:
Sorry. I missed this one. I can fix it.

mgencur commented on test/e2e/README.md:
```diff
@@ -1,17 +1,3 @@
 # End to end tests
 
 * [Running e2e tests](../README.md#running-e2e-tests)
```
Ok. Makes sense.

mgencur commented:
I signed it!

mgencur commented:
Hi Adriano, the rest of review comments should be addressed. Please let me know what you think. Thanks

adrcunha commented:
/approve

Martin, you have to resolve the merge conflict, so an LGTM now will be discarded by the next commit. Once you fix the merge I'll LGTM the PR.

adrcunha commented:
/approve
/lgtm

adrcunha commented:
Adding Matt for approving the updated to `/vendor`.

/assign mattmoor

grantr commented:
/approve

@mgencur can you fix the conflict? You can probably just back out your knative/pkg changes because master has moved past where you locked it.

mgencur commented:
Hi! I've fixed the conflicts. The Gopkg and vendor/ changes were in multiple commits and difficult to resolve. So I've merged all the commits into one, removed Gopkg and vendor/ changes. Then I did the rebase and added one more commit that resolves the remaining missing packages. Hope it's ok.

adrcunha commented on test/README.md:
```diff
@@ -14,26 +15,83 @@ Use `go test`:
-stack with Istio and run the end-to-end tests against the Knative Eventing built from source.
+## Running end-to-end tests
+
+To run [the e2e tests](./e2e) and [the conformance tests](./conformance), you need to have a running environment that meets
```
There are no conformance tests for eventing yet...

mgencur replied:
OK.

adrcunha commented on test/README.md:
````diff
@@ -14,26 +15,83 @@ Use `go test`:
+test/presubmit-tests.sh
+```
+
+_Note that to run `presubmit-tests.sh` or `e2e-tests.sh` scripts, you'll need kubernetes `kubetest` installed:_
````
I suggest linking to the "environment requirements" below instead of showing the `go get` command.

mgencur replied:
OK.

adrcunha commented on test/README.md:
```diff
@@ -1,10 +1,11 @@
+* [Unit tests](#running-unit-tests) reside in the codebase alongside the code they test
+* [End-to-end tests](#running-end-to-end-tests) reside in [`/test/e2e`](./e2e)
 
+If you want to add more tests, see [adding_tests.md](./adding_tests.md).
```
Isn't this on `e2e/README.md`?

mgencur replied:
No. But i'm removing this line because it references the adding_tests.md file which doesn't exist. See our earlier discussion on this topic: https://github.com/knative/eventing/pull/416#discussion_r217448958

mgencur commented:
Another day, another rebase. And a small fix to readme.

mgencur commented:
/test pull-knative-eventing-integration-tests

vaikas-google commented:
/retest

mgencur commented:
Yay!!!!
